src/java.base/share/classes/jdk/internal/foreign/abi/CallBufferCache.java

```diff
@@ -0,0 +1,114 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.foreign.abi;
+
+import jdk.internal.misc.TerminatingThreadLocal;
+import jdk.internal.misc.Unsafe;
+import jdk.internal.vm.Continuation;
+
+/**
+ * Provides carrier-thread-local storage for up to two buffer addresses.
+ * It is caller's responsibility to store homogeneous segment sizes.
+ * Storing addresses, not MemorySegments turns out to be slightly faster (write barrier?).
+ */
+public final class CallBufferCache {
+    private static final Unsafe UNSAFE = Unsafe.getUnsafe();
+
+    static class PerThread {
+        // Two-elements to support downcall + upcall.
+        private long address1;
+        private long address2;
+
+        long pop() {
+            if (address1 != 0) {
+                long result = address1;
+                address1 = 0;
+                return result;
+            }
+            if (address2 != 0) {
+                long result = address2;
+                address2 = 0;
+                return result;
+            }
+            return 0;
+        }
+
+        boolean push(long address) {
+            if (address1 == 0) {
+                address1 = address;
+                return true;
+            }
+            if (address2 == 0) {
+                address2 = address;
+                return true;
+            }
+            return false;
+        }
+
+        void free() {
+            if (address1 != 0) CallBufferCache.free(address1);
+            if (address2 != 0) CallBufferCache.free(address2);
+        }
+    }
+
+    @SuppressWarnings("restricted")
+    public static long allocate(long size) {
+        return UNSAFE.allocateMemory(size);
+    }
+
+    public static void free(long address) {
+        UNSAFE.freeMemory(address);
+    }
+
+    private static final TerminatingThreadLocal<PerThread> tl = new TerminatingThreadLocal<>() {
+        @Override
+        protected PerThread initialValue() {
+            return new PerThread();
+        }
+
+        @Override
+        protected void threadTerminated(PerThread cache) {
+            cache.free();
+        }
+    };
+
+    public static long acquire() {
+        // Protect against vthread unmount.
+        Continuation.pin();
```

Comment:
Other code that does this, first checks if we are in a virtual thread.
Note that we have two options for dealing with problematic cases:
* pin/unpin as soon as we see a virtual thread
* if we have a virtual thread and we detect a race in accessing the cache (e.g. because of how the virtual threads have been scheduled to the same underlying carrier thread), we could avoid pinning, but just allocate a fresh new segment (thus avoiding recycling issues)

Reply:
Would be worth understanding why we would _not_ want to pin here.
IINM, the `pin` code itself [checks](https://github.com/mernst-github/jdk/blob/f5573f5cbdcae5d1303c8b58d2946c168b977326/src/hotspot/share/opto/library_call.cpp#L3754) whether there's continuation, what would we win by a redundant `if` ?
Second, I'm not sure how you would want to detect a race, if you cannot trust any field you just read.
I can't see how you would get around using atomics, which is a much more expensive alternative.

I designed these to be as short sections as possible (if available, take/null out slot ;  if free, put slot; ). If that isn't what pinning is good for, I'm not sure what is.